[Merged by Bors] - feat(measure_theory/order/upper_lower): Order-connected sets in `ℝⁿ` are measurable

YaelDillies opened the pull request:
Prove that the frontier of an order-connected set in `ℝⁿ` (with the `∞`-metric, but it doesn't actually matter) has measure zero.

As a corollary, antichains in `ℝⁿ` have measure zero.

Co-authored-by: @JasonKYi

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #16955
- [x] depends on: #16956
- [x] depends on: #17257
- [x] depends on: #18629
- [x] depends on: #18635
- [x] depends on: #18636

This is a three lines remark in [*Reverse Kleitman Inequalities*, Bollobas, Leader, Radcliffe](https://londmathsoc.onlinelibrary.wiley.com/doi/epdf/10.1112/plms/s3-58.1.153), with no proof. So Jason, Bhavik and I came up with a proof using Lebesgue Density Theorem.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies committed:
initial commit

YaelDillies committed:
is_upper_set.exists_subset_ball

YaelDillies committed:
initial commit

YaelDillies committed:
done

YaelDillies committed:
whoops #lint

YaelDillies committed:
fix lemma names

YaelDillies committed:
upper_closure_inter_lower_closure

YaelDillies committed:
Merge remote-tracking branch 'origin/upper_closure' into ord_connected_measurable

YaelDillies committed:
bump upper_lower

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
stupid api

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
most of the glue

YaelDillies committed:
fix false statement

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
done 🎉

YaelDillies committed:
feat(topology/algebra/order/upper_lower): The closure of an upper set

YaelDillies committed:
module docs

sgouezel commented on src/measure_theory/order/upper_lower.lean:
```diff
@@ -0,0 +1,115 @@
+/-!
+# Order-connected sets are null-measurable
+
+This file proves that order-connected sets in `ℝⁿ` under the pointwise order are measurable.
```
Can you recall what order-connected means here? The concept is clear to me in linear orders, less clear in general orders. In particular, a few examples would help: the PR shows that the set is null-measurable because the boundary has measure zero, but is there an example showing that it could be non-measurable?

YaelDillies committed:
map of upper sets
upper sets form a monoid
upper set of open is open
antichains have measure zero

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
bump
more map API

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
bump

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into upper_lower_topology

YaelDillies committed:
thickening lemmas

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
bump

YaelDillies committed:
remove norm results

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into upper_lower_topology

YaelDillies committed:
revert analysis.normed.order.basic

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
Merge remote-tracking branch 'origin/upper_lower_topology' into ord_connected_measurable

YaelDillies committed:
split topological and norm results

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
bump

YaelDillies committed:
bdd_below.closure

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
bump

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
boundedness of neighborhoods

YaelDillies committed:
dedup lemmas

YaelDillies committed:
generalise data.set.finite lemmas

YaelDillies committed:
fix data.set.finite

YaelDillies committed:
unsorry

YaelDillies committed:
fix measure_theory.order.upper_lower

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
bump

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into ord_connected_measurable

YaelDillies committed:
move lemmas

jcommelin commented on src/analysis/normed/order/upper_lower.lean:
```diff
@@ -99,7 +110,78 @@ end
+(dist_mono_right h₁ (h₁.trans hb) hb).trans $
+  dist_anti_left (ha.trans $ h₁.trans hb) (h₁.trans hb) ha
+
+protected lemma metric.bounded.bdd_below : metric.bounded s → bdd_below s :=
```
```suggestion
protected lemma metric.bounded.bdd_below : bounded s → bdd_below s :=
```

jcommelin commented on src/measure_theory/order/upper_lower.lean:
```diff
@@ -0,0 +1,125 @@
+
+variables {ι : Type*} [fintype ι] {s : set (ι → ℝ)} {x y : ι → ℝ} {δ : ℝ}
+
+private lemma aux₀
```
Maybe add a short comment with a forward pointer, explaining where these aux lemmas are used.

jcommelin commented on src/measure_theory/order/upper_lower.lean:
```diff
@@ -0,0 +1,125 @@
+  all_goals { positivity },
+end
+
+private lemma aux₁
```
Idem

jcommelin reviewed (commented):
bors d+

YaelDillies committed:
forward pointer
